### ISSUE TITLE
Passing a filter statement to session when updating emits an update without any clauses

### ISSUE THREAD

**Issue by ramonvg** [sqlalchemy/sqlalchemy#4556]
Today we released to production a statement that looks like this:
```
session.query(User.name == 'filter').update({'name': 'update'})
```
as you can see, it's a wrong version of:
```
session.query(User).filter(User.name == 'filter').update({'name': 'update'})
```
and the emited SQL was the following, which updated the whole table.
```
UPDATE users SET name='update'
```

We would like to know if there's a reason to allow this behavior and if there's not, maybe it'd be a good idea to block this kind of statement. 

Working example:
```
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, Integer, String, create_engine
from sqlalchemy.orm import sessionmaker

engine = create_engine("postgresql://root:root@postgres/test")
Session = sessionmaker(bind=engine)
session = Session()
Base = declarative_base()


class User(Base):
    __tablename__ = 'users'
    id = Column(Integer, primary_key=True)
    name = Column(String(50))

Base.metadata.create_all(engine)

session.query(User.name == 'filter').all()
# SELECT users.name = 'filter' AS anon_1 FROM users

session.query(User.name == 'filter').update({'name': 'update'})
# UPDATE users SET name='update'
```

**Comment by zzzeek** [comment]
> 
> We would like to know if there's a reason to allow this behavior and if there's not, maybe it'd be a good idea to block this kind of statement.

this has been supported behavior mostly because it is valid to say something like, `query(User.__table__).update(...)`, when that form is used, it is equivalent to saying `query(user_table.c.id, user_table.c.name, user_table.c.<...>, ...).update()`, so we consider a Column expression to be valid.  your query was definitively against a certain table and the internals don't make it easy to figure out this was not your intent, although it is likely possible to guard against this *specific* case, sure.   

There is a long term tentative plan for SQLAlchemy 2 where Query() would no longer have an update() method at all and the regular update() construct would be used, so that there would be less awkwardness in this regard.

> Today we released to production a statement that looks like this:

I'm not sure if this is intended to suggest some form of urgency on SQLAlchemy's part to alter this behavior ; as mentioned above, we can always add logic to test for a specific mistake someone made, but that is just one out of potentially infinite other kinds of mistakes we didn't think of that day, so depending on the situation (this one is sort of in the middle) it may be more or less feasible as a generally useful catch that can prevent a large class of mistakes while not disallowing for valid use cases or breaking reasonable backwards compatibility (this one would break backwards compatibility for someone relying on this behavior right now which makes it less likely for anything to change soon).

It's of course imperative that code be tested before being put on production and the behavior here is deterministic, so proper testing would have prevented this mistake from causing damage.

**Comment by squeaky-pl** [comment]
Hi Mike, we are not trying to suggest some any urgency here. We were just little bit surprised by the SQL expression it compiles (as it's not immediately clear it would compile something that looks like Boolean comparison on Python side to this) but I get how sqlalchemy guesses from that column expression that table name and there are people relying on this behavior.

**Comment by ramonvg** [comment]
Hi again, that was clearly our mistake and definetely we should catch those bugs on testing. We love SQLAlchemy and we just want users not repeating our mistakes, that's basically why I was opening the issue.There was not any hostility intended. Sad to hear that you read it that way!

**Comment by zzzeek** [comment]
> Hi again, that was clearly our mistake and definetely we should catch those bugs on testing. We love SQLAlchemy and we just want users not repeating our mistakes, that's basically why I was opening the issue.There was not any hostility intended. Sad to hear that you read it that way!

Hi Ramon -

I would say that I sense frustration in that SQLAlchemy did something that screwed up your production database, but as far as "hostility" I don't perceive any hostility here, my own internal tendency towards projection (that is, my imagination of an emotion that then clouds my perception of an interpersonal situation) is very hyperactive and that's what you might have seen me tweeting about (that I need to meditate and try to figure out where my own reactions are clouding things), I try to get that under wraps before I actually respond to things and I apologize if I failed to cap it enough in this case.

as for this issue we can catch the mistake but it needs to be some kind of heuristic that is ensuring it doesn't block valid use cases while catching the bad ones; probably a flag that is set up when someone passes a Table object to Query will allow us to differentiate that kind of column expression from other arbitrary ones.

**Comment by sqla-tester** [comment]
**ramonvg** has proposed a fix for this issue in the **master** branch:

**Fixes: #4556 - Test that prevents passing a filter to Session.query when running an update** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2052
